chore: add ADR-0009 (#54)
helpwave/helpwave: development/adr/0009-versioning-strategy.md

# **ADR-009: Versioning strategy**

## Status

Proposed

## Context

As a software product, we should establish a sense of versioning.
With versioning we are able to pin our products as a versioned resource at customer facing deployments
or for feature planing. This ADR is focused on the Git implementation of our versioning strategy.

## Decision

The base of every versioning strategy is the version number. We are using [Semantic Versioning 2.0.0](https://semver.org/) with version prefix of `v`.

> Our Git strategy is based on the branching and versioning strategy of [Dapr](https://github.com/dapr/dapr).

Reviewer: Move the dapr line to the bottom somewhere, as a pointer for further reading.
When a new dev reads this, they first check out semver.org and then the dapr repo, without any direction on what to look for

> - New changes are made against the `main` branch.

Reviewer: ```suggestion
- New changes are pushed to the `main` branch.
```

Reviewer: or
```suggestion
- New changes are comitted to the `main` branch.
```

> - For every `MINOR` release, we are opening a branch called `release-v[MAJOR].[MINOR]` (release branch) based of the previous release.

Reviewer: ```suggestion
- For every `MINOR` release, we create a new branch `release-v[MAJOR].[MINOR]` (_release branch_) based on the previous release on the same `MAJOR` track.
```

- New changes for the release branch are either merged or cherry-picked from `main`.

> - The version of the software on a release branch, until tagged, is called pre-release.

Reviewer: `pre-release` already has a meaning in the semver spec, we should choose something else

Author: [The meaning of a `pre-release` is not clear IMO. Everything that is `[0-9A-Za-z-].` after the patch version seperation via `-` is a `pre-release`](https://semver.org/#spec-item-9).

But yes, we should define `pre-releases` more precisely. Can we even find out the upcoming version on the release branch just from new commits? If a commit follows the tag `v1.0.0` on the branch `release-v1.0`, is the newly created commit automatically the pre-release `v1.0.1-rc-[commit-hash]`? (rc = release candidate)

Author: I've added `Continues tagging of the pre-release commits is optional and is at the discretion of the particular implementation of the repository.`

> - The tag should be called `v[MAJOR].[MINOR].[PATCH]`. Based on the versioning number of SemVer.

Reviewer: ```suggestion
- When ready to ship, the latest commit on a release branch is tagged `v[MAJOR].[MINOR].[PATCH]` based on the version number.
```

- A tag of a commit on the release branch indicates a released version.

> - If a Bugfix needs to take place for all versions, the fix can be made against the `main` branch and cherry-picked into the release branches.

Reviewer: already explained in ln. 20

Author: This line addresses bug fixes, the other line features.

> - If a Bugfix needs to take place only for a specific version, this fix can be made against the release branch. 

Reviewer: ```suggestion
- If a change needs to be made only for a specific version, it can be made on the release branch 
```
but move it to line 21

> ![](../../assets/versioning-strategy-visualization.png)

Reviewer: add a heading and/or short description to the image

## Consequences

As long as our versioning strategy stays consistent across our whole product landscape.

> The major version is a compatibility identifier. `web/tasks/v1.1.1` is compatible with `services/v1.2.3`. 

Reviewer: ```suggestion
The major version is a compatibility identifier. `web/tasks/v1.1.1` is compatible with `services/v1.2.3`, as `MINOR` versions add functionality in a backward compatible manner.
```

# **ADR-0009: Versioning strategy**

## Status

Proposed

## Context

As a software product, we should establish a sense of versioning.
With versioning we are able to pin our products as a versioned resource at customer facing deployments
or for feature planing. This ADR is focused on the Git implementation of our versioning strategy.

## Decision

The base of every versioning strategy is the version number. We are using [Semantic Versioning 2.0.0](https://semver.org/) with version prefix of `v`.

- New changes are comitted to the `main` branch.
- For every `MINOR` release, we create a new branch `release-v[MAJOR].[MINOR]` (_release branch_) based on the previous release on the same `MAJOR` track.

> - New changes for the release branch are either merged or cherry-picked from `main`.

Author: Therefore, no explicit recommendation for merging or cherry-picking. We can not set a general procedure for the future.
